### ISSUE TITLE
Fail to infer instance member of Literal type

### ISSUE THREAD

**Issue by tamuhey** [microsoft/pyright#380]
```python
class A:
    def __init__(self, a: Literal["a"]):
        self.a = a

    def foo(self, a: Literal["a"]):
        pass

    def bar(self):
        self.foo(self.a) # Argument of type 'str' cannot be assigned to parameter 'a' of type 'Literal['a']' 'str' cannot be assigned to 'Literal['a']'
```

Is this behavior intended?

**Comment by erictraut** [comment]
This is the intended behavior. Pyright needs to make certain assumptions about when to strip the "Literal" from a type and when to keep it. When inferring types, it needs to be somewhat conservative here — especially for instance members of classes that can be assigned outside the scope of the module.

The heuristics that pyright uses to determine whether to strip the literal when inferring the type of instance members:
1. If it appears to be a constant (named as all-caps), don't strip the literal.
2. If it appears to be a private member (starts with underscore), don't strip the literal because all assignments to the private member are assumed to be within the class definition, and the analyzer can reason about all of them.
3. In all other cases, strip the literal. This allows assignments to the member from outside the class definition that don't honor the literal.

Without rule 3, a bunch of common patterns would be considered errors. For example,
```
class A:
    def __init__(self):
        # If literals weren't stripped here, enable_feature
        # would be inferred to be type Literal[False], and it
        # could never be set to True.
        self.enable_feature = False
```

So in the example above, if you rename your instance member from `a` to `_a`, you'll get the behavior you're looking for. Another solution is to provide an explicit type annotation (`self.a: Literal["a"] = a`).
